model/src/main/kotlin/licenses/LicenseInfoResolver.kt

```diff
@@ -203,28 +206,47 @@ class LicenseInfoResolver(
 
         val licenseInfo = resolveLicenseInfo(id)
         val licenseFiles = mutableListOf<ResolvedLicenseFile>()
-        val matcher = FileMatcher(archiver.patterns)
 
         licenseInfo.flatMapTo(mutableSetOf()) { resolvedLicense ->
             resolvedLicense.locations.map { it.provenance }
         }.forEach { provenance ->
             val archiveDir = createTempDirectory("$ORT_NAME-archive").toFile().apply { deleteOnExit() }
-
             val path = "${id.toPath()}/${provenance.hash()}"
-            if (archiver.unarchive(archiveDir, path)) {
-                archiveDir.walk().forEach { file ->
-                    val relativePath = file.toRelativeString(archiveDir)
-                    if (matcher.matches(relativePath)) {
-                        licenseFiles += ResolvedLicenseFile(
-                            provenance, licenseInfo.filter(provenance, relativePath), relativePath, file
-                        )
-                    }
-                }
+
+            if (!archiver.unarchive(archiveDir, path)) return@forEach
+
+            val rootLicenseFiles = getApplicableLicenseFilesForDirectory(
+                relativeFilePaths = archiveDir.walk().filter { it.isFile }.mapTo(mutableSetOf()) {
+                    it.relativeTo(archiveDir).toString()
+                },
+                directory = provenance.vcsInfo?.path.orEmpty()
+            )
+
+            licenseFiles += rootLicenseFiles.map { relativePath ->
+                ResolvedLicenseFile(
+                    provenance = provenance,
+                    licenseInfo.filter(provenance, relativePath),
+                    relativePath,
+                    archiveDir.resolve(relativePath)
+                )
             }
         }
 
         return ResolvedLicenseFileInfo(id, licenseFiles)
     }
+
+    private fun getApplicableLicenseFilesForDirectory(
+        relativeFilePaths: Collection<String>,
+        directory: String
+    ): Set<String> {
```

Comment:
What do you think of adding this function to the `RootLicenseMatcher` instead? And maybe later generalize the implementation somehow so that the fake findings are not required?

Reply:
Would you be fine with a TODO for that? I thought as well that it should be refactored, but I would like to do further changes before doing so - so that I can learn more to avoid some back and forth refactoring.

Reply:
ok, I changed it already in this PR.